clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -369,6 +369,71 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
         }
     }
 
+    /**
+     * Compute size in bytes of record body and size in bytes of a full record without depressed.
+     *
+     * @param input
+     * @param baseOffset
+     * @param baseTimestamp
+     * @param baseSequence
+     * @param logAppendTime
+     * @return
+     * @throws IOException
+     */
+    public static SimplifiedDefaultRecord simplifiedreadFrom(DataInput input,
+                                                             long baseOffset,
+                                                             long baseTimestamp,
+                                                             int baseSequence,
+                                                             Long logAppendTime) throws IOException {
+        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
+        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
+        return simplifiedreadFrom(input, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
+                baseSequence, logAppendTime);
+    }
+
+    private static SimplifiedDefaultRecord simplifiedreadFrom(DataInput input,
+                                                              int sizeInBytes,
+                                                              int sizeOfBodyInBytes,
+                                                              long baseOffset,
+                                                              long baseTimestamp,
+                                                              int baseSequence,
+                                                              Long logAppendTime) throws IOException {
+        try {
+            byte attributes = input.readByte();
+            int skipBytes = 1;
+            long timestampDelta = ByteUtils.readVarlong(input);
+            long timestamp = baseTimestamp + timestampDelta;
+            if (logAppendTime != null)
+                timestamp = logAppendTime;
+            skipBytes += ByteUtils.sizeOfVarlong(timestampDelta);
+
+            int offsetDelta = ByteUtils.readVarint(input);
+            long offset = baseOffset + offsetDelta;
+            int sequence = baseSequence >= 0 ?
+                    DefaultRecordBatch.incrementSequence(baseSequence, offsetDelta) :
+                    RecordBatch.NO_SEQUENCE;
+            skipBytes += ByteUtils.sizeOfVarlong(offsetDelta);
+
+            int keySize = ByteUtils.readVarint(input);
+            boolean hasKey = false;
+            if (keySize >= 0) {
+                hasKey = true;
+            }
+            skipBytes += ByteUtils.sizeOfVarlong(keySize);
+
+            skipBytes = sizeOfBodyInBytes - skipBytes;
+            if (skipBytes > 0) {
+                int currentSkipBytes = input.skipBytes(skipBytes);
```

Comment:
I think the proposal makes sense since under the mentioned conditions we only validate:

1) hasKey if compacted topic
2) timestamp within max_diff range.
3) offset monotonically increasing hence inPlaceAssignment is possible.

All these does not require key/value content actually, so leveraging on `skipBytes` to bypass the compressed key / value sounds reasonable.

Of course it means that if the broker ever needs to add any functionality in the future that rely on exposing the `Record` in full, this savings needs to be returned back again.

Reply:
Thanks for your support and suggestion for this PR.

Reply:
`skipBytes` doesn't avoid decompression though, the `readBlock` call below decompresses the buffer:

```java
class: KafkaLZ4BlockInputStream

public long skip(long n) throws IOException {
        if (finished) {
            return 0;
        }
        if (available() == 0) {
            readBlock();
        }
        if (finished) {
            return 0;
        }
        int skipped = (int) Math.min(n, available());
        decompressedBuffer.position(decompressedBuffer.position() + skipped);
        return skipped;
    }
```

We do avoid GC pressure though.

Reply:
According to the result of CPU reduction, the skip operation does not reduce data decompression, but it reduces CPU consumption. Is a large amount of CPU consumption in the GC pressure？

Reply:
**class:DefaultRecord**
1. Function simplifiedreadFrom():
 public static SimplifiedDefaultRecord **simplifiedreadFrom**(DataInput input,
                                                             long baseOffset,
                                                             long baseTimestamp,
                                                             int baseSequence,
                                                             Long logAppendTime) throws IOException {
        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
        return simplifiedreadFrom(input, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
                baseSequence, logAppendTime);
    }
1. Function readFrom():
    public static DefaultRecord **readFrom**(DataInput input,
                                         long baseOffset,
                                         long baseTimestamp,
                                         int baseSequence,
                                         Long logAppendTime) throws IOException {
        int sizeOfBodyInBytes = ByteUtils.readVarint(input);
        ByteBuffer recordBuffer = ByteBuffer.allocate(sizeOfBodyInBytes);
        **input.readFully(recordBuffer.array(), 0, sizeOfBodyInBytes);**
        int totalSizeInBytes = ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes;
        return readFrom(recordBuffer, totalSizeInBytes, sizeOfBodyInBytes, baseOffset, baseTimestamp,
                baseSequence, logAppendTime);
    }

Reply:
We removed "**input.readFully(recordBuffer.array(), 0, sizeOfBodyInBytes);**" at  function **simplifiedreadFrom**() . So we do not call function **readBlock()** at **KafkaLZ4BlockInputStream** call.So,we don't decompress buffered data.

Reply:
> I pasted the code from `skip` which is called from `skipBytes`. I know we don't call `readFully`.
I understand what you mean.I will check again and do performance testing again.And we will find the reason why it reduces CPU consumption after modified code.Thanks.

Reply:
I understand what you mean.I will check again and do performance testing again.And we will find the reason why it reduces CPU consumption after modified code.Thanks .

Reply:
@ijuma 
After we checked and completed the performance test again, we located the code "**ByteBuffer recordBuffer = ByteBuffer.allocate(sizeOfBodyInBytes);**(**Class:DefaultRecord,Function:readFrom**)” which consumed CPU resources and caused a lot of GC  .Our modified code reduces the allocation and copying of ByteBuffer, so the test performance is greatly improved, and the CPU's stable usage is **below 60%**. The following is a comparison of different code test performance under the same conditions.

**Result of performance testing** 
Main config of Kafka: Single Message:1024B;TopicPartitions:200;linger.ms:1000ms,
1.**Before** modified code(Source code):
Network inflow rate:600M/s;**CPU(%)(97%);production:25,000,000 messages/s**
2.**After** modified code(remove allocation of ByteBuffer):
Network inflow rate:1G/s;**CPU(%)(<60%);production:41,000,000 messages/s**

Reply:
**1.Before modified code(Source code) GC:**
![](https://i.loli.net/2019/05/07/5cd16df163ad3.png)
**2.After modified code(remove allocation of ByteBuffer) GC:**
![](https://i.loli.net/2019/05/07/5cd16dae1dbc2.png)

Reply:
That's what I thought, thanks for confirming. Can you please update the PR and JIRA description to reflect this?

Reply:
ok ,I will target trunk create PR  and create a new issue.Thanks

Reply:
The new PR : https://github.com/apache/kafka/pull/6699